### DISCUSSION TITLE
How to access the router from an imported TS file?

### DISCUSSION THREAD

**Discussion by wsw70** [quasarframework/quasar#12569] [General - Components / Directives / etc] [answered]
I have a `helpers.ts` file that I import in my components. In this file, I would like to export a function:

```
// goToCase changes the URL to the case Uuid
const goToCase = (uuid: string) => {
  router.push(`${frontendUrl}/${uuid}`)
}
```

**My problem**: I do not know what to import `router` from so that I can do [programmatic navigation](https://router.vuejs.org/guide/essentials/navigation.html#programmatic-navigation)

**Question 1**: should I do programmatic navigation the Vue way, or is there a specific Quasar way to do it (I did not fond anything in the docs)

**Question 2**: if so - where can I import `router` from (in components and helper files)

**Comment by MilosPaunovic** [comment]
You could move your `helpers.ts` in `boot` folder, register it as a boot file in `quasar.conf.js` and in there, you'll have access to router.

```js
export default ({ router }) => {
  // something to do
}
```

For more info, take a look at this:
[https://quasar.dev/quasar-cli/boot-files](https://quasar.dev/quasar-cli/boot-files)

**Reply by wsw70** [reply]
Thank you. I am trying to wrap my head around the intended use.

I declared a file `setUrl.ts` in `quasar.conf.js` (it is properly registered)

In this file, I tried to create a function that would take a parameter (`uuid`) and do something with it:

```
export default ({ router }) => {
   router.push(`${uuid}`)
}
```

I also tried incantations of the form

```
export default ({ router }) => {
  const setUrl = (uuid: string) => {
    router.push(`${uuid}`)
  }
}
```


**What is the proper structure of this boot function**

(since I use TS I wanted to wrap the function with `boot` per the documentation but that will be the next step)

**Comment by yusufkandemir** [comment, marked as answer]
The best way will accept the router as the parameter, and assign it a default value to simplify the usage.

```ts
// Helper file
import { useRouter } from 'vue-router'

// ...

const goToCase = (uuid: string, router = useRouter()) => {
  router.push(`${frontendUrl}/${uuid}`)
}

// Inside <script setup> or setup()
goToCase('whatever')

// Inside boot files, router/index.js or anywhere you don't initiate the calling chain from setup()
// i.e. where provide/inject is not available
export default boot(({ router }) => {
  goToCase('whatever', router)
})
```

**Reply by wsw70** [reply]
Thank you. At some point during my experiments I arrived to something similar (= without using a boot file) but my problem was (and still is) that `Router()` returns `undefined`, and this breaks at `Router().push()`.

I had

```
export const goToCase = (uuid: string) => {
  void useRouter().push(`${frontendUrl}/${uuid}`)
}
```

as the function, but in your version `router` is `undefined` as well.

**Reply by yusufkandemir** [reply]
I don't know which `useRouter` you are using, or where you are calling `goToCase` from. So, I can't say much. If you can provide a full code example or a repro, I can help. It might also be helpful to know that you can't call inject inside event listeners, so something like this won't work too:

```html
<q-btn @click="onClick" />
```

```js
function onClick () {
  // This won't work !!!
  const router = useRouter()

  router.push('...')
}

// But, this will work !!!
const router = useRouter()
function onClick () {
  router.push('...')
}

// goToCase helper example that will work inside event listeners
const router = useRouter()
function onClick () {
  goToCase('...', router)
}
```

**Reply by wsw70** [reply]
Thank you - I just found out that moving `useRouter` outside the method like in your second case works. 

Thanks again - this is something that I really had a hard time understanding from the docs. But no what I know how to do it, the ones at https://router.vuejs.org/guide/advanced/composition-api.html#vue-router-and-the-composition-api seem obvious and I do not know how I could have missed them.

**Reply by yusufkandemir** [reply]
You are welcome. You might also consider using [named routes](https://router.vuejs.org/guide/essentials/named-routes.html#named-routes) to simplify your helper, or maybe remove the need for it. I hope it helps to simplify some stuff.
